src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -241,19 +253,215 @@ uct_ib_mlx5_devx_reg_ksm_data(uct_ib_mlx5_md_t *md,
+{
+    const size_t reg_length = UCT_IB_MD_MAX_MR_SIZE;
+    intptr_t addr           = (intptr_t)mr->super.ib->addr & ~(reg_length - 1);
+    /* FW requires indirect atomic MR addr and length to be aligned
+     * to max supported atomic argument size */
+    size_t length           = ucs_align_up(mr->super.ib->length +
+                              (intptr_t)mr->super.ib->addr - addr,
+                              md->super.dev.atomic_align);
+    int list_size           = ucs_div_round_up(length, reg_length);
+    int i;
+    char *in;
+    void *klm;
+    ucs_status_t status;
+
+    status = uct_ib_mlx5_alloc_mkey_inbox(list_size, &in);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    klm = UCT_IB_MLX5DV_ADDR_OF(create_mkey_in, in, klm_pas_mtt);
+    for (i = 0; i < list_size; i++) {
+        if (i == list_size - 1) {
+            UCT_IB_MLX5DV_SET(klm, klm, byte_count, length % reg_length);
+        } else {
+            UCT_IB_MLX5DV_SET(klm, klm, byte_count, reg_length);
+        }
+        UCT_IB_MLX5DV_SET(klm, klm, mkey, mr->super.ib->lkey);
+        UCT_IB_MLX5DV_SET64(klm, klm, address, addr + (i * reg_length));
+        klm = UCS_PTR_BYTE_OFFSET(klm, UCT_IB_MLX5DV_ST_SZ_BYTES(klm));
+    }
+
+    status = uct_ib_mlx5_devx_reg_ksm(md, addr + off, length, list_size,
+                                      reg_length, in, mr_p, mkey);
+    ucs_free(in);
+    return status;
+}
+
+static void uct_ib_mlx5_devx_md_mr_lru_pop(uct_ib_mlx5_md_t *md)
+{
+    uct_ib_mlx5_mem_lru_entry_t *head =
+            ucs_list_extract_head(&md->lru.rkeys_list,
+                                  uct_ib_mlx5_mem_lru_entry_t, link);
+    khint_t iter;
+
+    ucs_assert(kh_size(&md->lru.rkeys_hash) > 0);
+
+    ucs_trace("mr lru size=%d pop: %x", kh_size(&md->lru.rkeys_hash),
+              head->rkey);
+
+    iter = kh_get(rkeys, &md->lru.rkeys_hash, head->rkey);
+    ucs_assert(iter != kh_end(&md->lru.rkeys_hash));
+    kh_del(rkeys, &md->lru.rkeys_hash, iter);
+    ucs_free(head);
+}
+
+static ucs_status_t
+uct_ib_md_mlx5_devx_mr_lru_push(uct_ib_mlx5_md_t *md, uint32_t rkey)
+{
+    static size_t counter              = 0;
+    uct_ib_mlx5_mem_lru_entry_t *entry = ucs_malloc(sizeof(*entry),
+                                                    "rkey_lru_entry");
+    ucs_kh_put_t res;
+
+    if (entry == NULL) {
+        ucs_error("Cannot allocate rkey LRU entry");
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    entry->rkey = rkey;
+    ucs_list_add_tail(&md->lru.rkeys_list, &entry->link);
+    kh_put(rkeys, &md->lru.rkeys_hash, rkey, &res);
+    ucs_assert((res == UCS_KH_PUT_BUCKET_EMPTY) ||
+               (res == UCS_KH_PUT_BUCKET_CLEAR));
+    if (kh_size(&md->lru.rkeys_hash) > md->super.config.max_mr_gc_queue) {
+        uct_ib_mlx5_devx_md_mr_lru_pop(md);
+    }
+
+    if ((++counter % (md->super.config.max_mr_gc_queue)) == 0) {
+        md->mkey_tag += 1;
+        md->mkey_tag %= 251;
```

Comment:
`md->mkey_tag = (md->mkey_tag + 1) % 251;` /* Modulo prime number to avoid resonance with mkey */

Reply:
can we somehow document `251` (e.g. add constant)?
is it last prime number in [0..255] range? why it should be prime?